crypto/pkcs12/p12_crt.c

```diff
@@ -165,6 +168,19 @@ PKCS12_SAFEBAG *PKCS12_add_cert(STACK_OF(PKCS12_SAFEBAG) **pbags, X509 *cert)
     if (keyid && !PKCS12_add_localkeyid(bag, keyid, keyidlen))
         goto err;
 
+    /*
+     * Use trust set on certificate, if present, to set the Oracle trust bag attribute
+     */
+    trusts = X509_get0_trust_objects(cert);
+    if (trusts && sk_ASN1_OBJECT_num(trusts) > 0) {
+        if (sk_ASN1_OBJECT_num(trusts) > 1)
+            goto err;
+
+        trust = sk_ASN1_OBJECT_value(trusts, 0);
+        attr = X509_ATTRIBUTE_create(NID_oracle_jdk_trustedkeyusage, V_ASN1_OBJECT, trust);
+        X509at_add1_attr(&bag->attrib, attr);
+    }
+
```

Comment:
So this adds the attribute by default for anyone that has a trust setting on the cert supplied in the param. I'm not sure we want that? There may be trust settings on a cert because it has OpenSSL trust markings. I'm not sure we necessarily want to automatically publish that as a jdk trust marking?

I was expecting that the pkcs12 app would be responsible for adding this attribute. Although looking at how that app creates a PKCS12 object its not immediately clear how to do that. I would have expected the pkcs12 app to build up the set of PKCS12_SAFEBAG objects (adding the attribute if appropriate) and then create the PKCS12 object from that. However, although there are functions to manipulate PKCS12_SAFEBAGs, there doesn't seem to be an easy way to create a PKCS12 object from them. Are there missing API functions here?

Reply:
Very good question, I agree that it should not add by default unless there is an option set in the `openssl.cnf`.

Reply:
> Very good question, I agree that it should not add by default unless there is an option set in the openssl.cnf.

Right - and this change is currently changing the behaviour of the public API function `PKCS12_add_cert` in libcrypto. So it would impact any 3rd party application using libcrypto that happens to call that function. I would not expect such a function itself to be affected by openssl.cnf. The pkcs12 *app* should be affected by it - and only add the jdk attribute if the pkcs12 *app* entry in the config file says so.

Reply:
@mattcaswell I think the answer is simple -- is there other public API in libcrypto that is affected by openssl.cnf? If not, then this one should not be affected either.

But, we may consider introducing some API that would allow adding this attribute programmatically too.

Reply:
There are a handful. Mostly around provider and engine management. There are also some which affect libssl. But typically public API functions are not impacted by openssl.cnf.

Yes, we could add new API to enable adding the attributes.

Reply:
I'm not 100% sure I'm following

Reply:
> I'm not 100% sure I'm following

Ultimately this change is modifying the default behaviour of `PKCS12_add_cert` in such a way that it may behave unexpectedly for users that don't want the new functionality but are using OpenSSL's trust marking mechanism. We don't want that.

I actually think it is better to provide improved support for adding attributes to PKCS12_SAFEBAG objects in libcrypto - and leave it to the pkcs12 app to handle anything specific to jdkTrustedKeyUsage.

The problem is that the `PKCS12_create()`/`PKCS12_create_ex()` functions just take a set of certs and a key and generates a PKCS12 file from that. Internally it generates all the `PKCS12_SAFEBAG` objects but doesn't give you any opportunity to modify them by adding attributes.

I might envisage a `PKCS12_create_ex2()` function which does the same thing as `PKCS12_create_ex` but additionally takes a callback argument, e.g. something like this:

```` c
typedef int PKCS12_create_cb(PKCS12_SAFEBAG *bag, void *cbarg);

PKCS12 *PKCS12_create_ex2(const char *pass, const char *name, EVP_PKEY *pkey,
                         X509 *cert, STACK_OF(X509) *ca, int nid_key, int nid_cert,
                         int iter, int mac_iter, int keytype,
                         OSSL_LIB_CTX *ctx, const char *propq,
                         PKCS12_create_cb *cb, void *cbarg);
````

Everytime `PKCS12_create_ex2` creates a `PKCS12_SAFEBAG` to add to the PKCS12 file, it would call the callback. This would give the callback the opportunity to modify the `PKCS12_SAFEBAG` (for example by adding attributes).

The pkcs12 *app* could then use this new function to implement its own callback. The callback could then use `PKCS12_add1_attr_by_NID` to add arbitrary attributes to the `PKCS12_SAFEBAG`.

Reply:
@mattcaswell

Would this perhaps be a good opportunity to refactor both `PKCS12_create()` and `PKCS12_create_ex()` in such a way that they do not create all `PKCS12_SAFEBAG` objects internally, but call a new API (private or public) which can accept either bag attribute parameters or `NULL` if none are needed?

Then he could call that same API from `PKCS12_create()`, `PKCS12_create_ex()` and pass `NULL` and from `PKCS12_create_ex2` with the callback as you suggested passing actual bag attributes.

Basically, avoid the duplication of PKCS12 file structure creation?

Reply:
I'm open to concrete suggestions about what such an API would look like. My suggestion about `PKCS12_create_ex2` was intended to provide the flexibility to enable attributes to be associated with PKCS12_SAFEBAG objects easily. But its not the only way to do it. You could break up the implementation of the existing `PKCS12_create_ex` into smaller parts so that the construction of the `PKCS12_SAFEBAG` objects is separate to the construction of the PKCS12 object - and then expose some of those smaller parts as  API functions.

So, I'd be happy to see other ideas around how  to solve this problem.

Reply:
@mattcaswell 
> My suggestion about `PKCS12_create_ex2` was intended to provide the flexibility to enable attributes to be associated with `PKCS12_SAFEBAG` objects easily.

Your suggestion sounds OK to me except that it makes me cringe at the number of API parameters. Anything more than about 7 parameters in a public API is way too much and IMO those parameters should be passed as some sort of a context structure (by value or by reference) or even better the API should be split into smaller functional parts such as:

1. Creating a PKCS12 context, setting master password, encryption parameters, and file name
2. Creating an empty list of safebags associated with the above context
3. Creating a safebag (for shrouded keybag set encryption parameters, for cert bag attributes, also friendlyName/localKeyId where needed)
4. Adding such newly created safebag to the list of safebags, repeating from step 3 until done
5. Serializing the safebag list to PKCS12 context and saveing it to a file

That's a high-level breakdown of how I would implement the process if I were writing it from scratch. Whether it would be possible to fit this to the current API and how is of course a totally different question.

Reply:
I was thinking keeping `PKCS12_create_ex` as it is and adding another public API to specify the call back and optional argument. 

The new public API to set up the call back sets some private function pointer in the pkcs12 app and it's NULL by default to uses of the existing API won't be affected.

Reply:
Splitting the existing code into several sounds a good idea but not sure I want to do that under this PR. 

It needs a lot of thought about how we handle situations where users miss calling one of the new functions or calls them in the wrong order etc

Reply:
@grahamwoodward 

>Splitting the existing code into several sounds a good idea but not sure I want to do that under this PR.

Refactoring of existing public APIs to break out the functionality that would be common for old and new API in my opinion most certainly deserves a separate PR, maybe even a separate issue with some triage to better define how it should be restructured to better support this feature.

@mattcaswell 

Do you also think that would be the right way to proceed? If not, please advise.

Reply:
> Anything more than about 7 parameters in a public API is way too much

Yes, I agree the current PKCS12_create_ex API is less than ideal and splitting it up might not be a bad idea (although that is significantly more work and considerable scope creep from the original idea behind this PR).

> and IMO those parameters should be passed as some sort of a context structure (by value or by reference) or even better the API should be split into smaller functional parts such as:

Having a context structure would definitely be beneficial and enable things to be split up more easily. But again it is a considerably larger change than originally conceived.

> I was thinking keeping PKCS12_create_ex as it is and adding another public API to specify the call back and optional argument.

I don't see how that could work. Since there is no context structure currently its difficult to see where you would store the callback for future reference by `PKCS12_create_ex`. Unless of course you put it in the libctx or just make it global - but that would then impact *all* calls to `PKCS12_create_ex` which I don't think is desirable.

> Refactoring of existing public APIs to break out the functionality that would be common for old and new API in my opinion most certainly deserves a separate PR, maybe even a separate issue with some triage to better define how it should be restructured to better support this feature.

I am fine conceptually with refactoring the existing APIs. But if we go down this route then probably this will need some kind of a design document describing what the new APIs will look like and will need OTC (OpenSSL Technical Committee) approval. Here's an example design document that was produced recently:

https://github.com/openssl/openssl/blob/master/doc/designs/quic-design/dgram-api.md

Reply:
> I don't see how that could work. Since there is no context structure currently its difficult to see where you would store the callback for future reference by PKCS12_create_ex. Unless of course you put it in the libctx or just make it global - but that would then impact all calls to PKCS12_create_ex which I don't think is desirable.

I'm simply (and maybe this is wrong?) storing the callback within the p12_crt.c as a global. Then the places within PKCS12_create_ex in p12_crt which call PKCS12_add_cert() then check whether this callback isn't NULL and call it. The function itself is defined within the app.

If this isn't right, I'm happy to change it as you see fit.

Reply:
I'm fine to add the call back as a new argument to the PKCS12_create_ex (albeit a new PKCS12_create_ex2) and look to refactor that in a subsequent PR

Reply:
also @mattcaswell what are you envisaging is being passed in the void *cbarg? As far as I can see I only need the bag, which we're passing as the first arg to the callback. In my cb I'm doing all the work to read the config and set the attr (although not quite working, see my comment below)

Reply:
The `cbarg` is application defined and is passed back to the callback. It can be useful for passing contextual information from the application back to its own callback. This is something we do for most of our callbacks. Your particular implementation might not need it, but I think it is still good to have it.

> storing the callback within the p12_crt.c as a global

I don't think this is a good idea. Globals should be avoided. I think we either have to live with the excessively long param list to a new function `PKCS12_create_ex2` - or we have to do the refactor and split the API up (adding a context variable). I don't see a middle ground at the moment.

Reply:
@grahamwoodward 

> I'm simply (and maybe this is wrong?) storing the callback within the p12_crt.c as a global. Then the places within PKCS12_create_ex in p12_crt which call PKCS12_add_cert() then check whether this callback isn't `NULL` and call it. The function itself is defined within the app.

That doesn't sound thread-safe at all.

What guarantee do you have that someone on a different thread won't call `SetCallback(NULL)` after you test your global point, but before your thread gets the chance to dereference the pointer and execute the code it points to?

For that to work reliably you would have to introduce a dynamically allocated context in which, among other things, callback function pointer would be stored. Said context would be returned to the caller and they would have to pass it back to you on successive calls to your API so you know you are calling the right callback for the right tread / process.

As @mattcaswell said, doing this right would creep out of scope of this PR by a large margin. Regardless, I believe proper design should be made for this and reviewed by the OTC, because on projects as big and as widely used as OpenSSL a quick hack to add a relatively minor feature is a recipe for disaster.

Reply:
@grahamwoodward 

> also what are you envisaging is being passed in the void *cbarg?

The easiest way to imagine what `*cbarg` could be used for would be if your callback is set to a C++ class member function you could pass `this` pointer as `cbarg` so the member function can access private members of the class it belongs to.

It is the same pattern used when you create thread -- `void *param` is what gets passed to newly created thread function.

Reply:
yeah I get the idea as to why, we do the same for the code I work on when we have async call backs we'll allow a context to be passed so the callback can access stuff etc...I just meant in the context of this actually PR what he was envisaging but as matt said, there's probably nothing for my specific case but future cases might need something passed